File: guest-js/logger.ts
// Centralized logger for the serial plugin
// Avoids circular dependencies and code duplication

export const LogLevel = {
  None: 'None',
  Error: 'Error',
  Warn: 'Warn',
  Info: 'Info',
  Debug: 'Debug',
} as const;

export type LogLevel = (typeof LogLevel)[keyof typeof LogLevel];

// Global log level state
let currentLogLevel: LogLevel = LogLevel.Info;

/**
 * Sets the global log level
 */
export function setLogLevel(level: LogLevel): void {
  currentLogLevel = level;
}

/**
 * Gets the current log level
 */
export function getLogLevel(): LogLevel {
  return currentLogLevel;
}

/**
 * Checks if a message should be logged based on current log level
 */
function shouldLog(level: LogLevel): boolean {
  const levels: LogLevel[] = [
    LogLevel.None,
    LogLevel.Error,
    LogLevel.Warn,
    LogLevel.Info,
    LogLevel.Debug,
  ];
  const currentIndex = levels.indexOf(currentLogLevel);
  const requestedIndex = levels.indexOf(level);
  return requestedIndex <= currentIndex && currentIndex > 0;
}

/**
 * Logs an error message
 */
export function logError(...args: any[]): void {
  if (shouldLog(LogLevel.Error)) {
    console.error(...args);
  }
}

/**
 * Logs a warning message
 */
export function logWarn(...args: any[]): void {
  if (shouldLog(LogLevel.Warn)) {
    console.warn(...args);
  }
}

/**
 * Logs an info message
 */
export function logInfo(...args: any[]): void {
  if (shouldLog(LogLevel.Info)) {
    console.log(...args);
  }
}

/**
 * Logs a debug message
 */
export function logDebug(...args: any[]): void {
  if (shouldLog(LogLevel.Debug)) {
    console.log(...args);
  }
}

File: src/logger.rs
//! Centralized logging module for the serial plugin
//! 
//! This module provides a unified logging interface with configurable log levels.
//! All logging in the plugin should use these macros to respect the global log level setting.

/// Logs an error message if the current log level permits
#[macro_export]
macro_rules! log_error {
    ($($arg:tt)*) => {
        if $crate::state::get_log_level().should_log_error() {
            eprintln!($($arg)*);
        }
    };
}

/// Logs a warning message if the current log level permits
#[macro_export]
macro_rules! log_warn {
    ($($arg:tt)*) => {
        if $crate::state::get_log_level().should_log_warn() {
            println!($($arg)*);
        }
    };
}

/// Logs an info message if the current log level permits
#[macro_export]
macro_rules! log_info {
    ($($arg:tt)*) => {
        if $crate::state::get_log_level().should_log_info() {
            println!($($arg)*);
        }
    };
}

/// Logs a debug message if the current log level permits
#[macro_export]
macro_rules! log_debug {
    ($($arg:tt)*) => {
        if $crate::state::get_log_level().should_log_debug() {
            println!($($arg)*);
        }
    };
}

#[cfg(test)]
mod tests {
    use crate::state::{set_log_level, LogLevel};

    #[test]
    fn test_log_level_none() {
        set_log_level(LogLevel::None);
        // These should not panic, just not print anything
        log_error!("This should not be printed");
        log_warn!("This should not be printed");
        log_info!("This should not be printed");
        log_debug!("This should not be printed");
    }

    #[test]
    fn test_log_level_error() {
        set_log_level(LogLevel::Error);
        log_error!("Error message");
        // Warn, Info, Debug should not print
    }

    #[test]
    fn test_log_level_debug() {
        set_log_level(LogLevel::Debug);
        log_error!("Error message");
        log_warn!("Warning message");
        log_info!("Info message");
        log_debug!("Debug message");
    }
}

